relay/DCUtR.md

```diff
@@ -0,0 +1,158 @@
+connection, in which case `B` attempts a unilateral connection upgrade
+by initiating a direct connection to `A`.
+
+If the unilateral connection upgrade attempt fails or if `A` is itself a NATed
+peer that doesn't advertise public address, then `B` initiates the direct
+connection upgrade protocol as follows:
+1. `B` opens a stream to `A` using the `/libp2p/dcutr` protocol.
+2. `B` sends to `A` a `Connect` message containing its observed (and possibly
+   predicted) addresses from identify and starts a timer to measure RTT of the
+   relay connection.
+3. Upon receving the `Connect`, `A` responds back with a `Connect` message
+   containing its observed (and possibly predicted) addresses.
+4. Upon receiving the `Connect`, `B` sends a `Sync` message and starts a timer
+   for half the RTT measured from the time between sending the initial `Connect`
+   and receiving the response.
+5. Simultaneous Connect
+   - Upon receiving the `Sync`, `A` immediately starts a direct dial to B using
+     the addresses obtained from the `Connect` message.
+   - Upon expiry of the timer, `B` starts a direct dial to `A` using the
+     addresses obtained from the `Connect` message.
+6. On failure go back to step (2), reusing the same stream opened in (1).
+   Inbound peers (here `B`) SHOULD retry twice (thus a total of 3 attempts)
+   before considering the upgrade as failed.
+
+The purpose of the `Sync` message and `B`'s timer is to allow the two peers to
+synchronize so that they perform a simultaneous open that allows hole punching
+to succeed.
+
+If the direct connection is successful, then the peers should migrate
+to it by prioritizing over the existing relay connection. All new
+streams should be opened in the direct connection, while the relay
+connection should be closed after a grace period.  Existing indefinite
+duration streams will have to be recreated in the new connection once
+the relay connection is closed.
+
+
+### RPC messages
+
+All RPC messages sent over a stream are prefixed with the message length in
+bytes, encoded as an unsigned variable length integer as defined by the
+[multiformats unsigned-varint spec][uvarint-spec].
+
+Implemntations SHOULD refuse encoded RPC messages (length prefix excluded)
+larger than 4 KiB.
+
+RPC messages conform to the following protobuf schema:
+
+```proto
+syntax = "proto2";
+
+package holepunch.pb;
+
+message HolePunch {
+  enum Type {
+    CONNECT = 100;
+    SYNC = 300;
+  }
+
+  optional Type type=1;
+
+  repeated bytes ObsAddrs = 2;
+}
+```
+
+## FAQ
+
+- *Why exchange `CONNECT` and `SYNC` messages once more on each retry?*
+
+  Doing an additional CONNECT and SYNC for each retry prevents a flawed RTT
+  measurement on the first attempt to distort all following retry attempts.
+
+- *Why reuse the same stream for retries?*
+
+  Stream opening and stream protocol negotiation might distort the measured
+  round-trip-time. Reusing the stream from the first attempt allows cutting out
```

Comment:
I'm not sure I understand how. The RTT is determined as the time difference between the sent and received CONNECT messages. You can only send this message after the stream protocol has been negotiated. In case you pipeline this message, you'll only get a CONNECT in response if the protocol was accepted, so there shouldn't be any delay (at least no delay O(RTT)).
Are we worried about (probably sub-ms) processing delays introduced by multistream? If we are, should we also be worried about the time it takes to assemble the addresses to put in the CONNECT message, as well as the serialization delay of the protobuf?

Reply:
The above makes no assumptions about the multiplexer and stream negotiation protocol in use.

E.g. say that a stream multiplexer needs a whole round-trip to open a new stream. If the negotiation protocol uses pipelining, but does not wait for the stream to be opened before accepting data, the measurement would be distorted.

I acknowledge that this example seems contrived, that it might be fine making the assumption and thus that there is no need for reusing the same stream.

> Are we worried about (probably sub-ms) processing delays introduced by multistream?

I don't think we need to care about same-machine processing delays. At least not at this point in time without proper measurements.

I am fine either way. I like the current version, not making any assumptions about the multiplexer and protocol negotiation protocol. I like the simplicity of opening a new stream on each attempt. With the additional context above, what do you prefer @marten-seemann?

Reply:
No really strong opinion, but I think I'd prefer to do use separate streams. It would simplify timeout handling, and allow peers to wait for a certain time between retries without keeping the stream open.

Reply:
:+1: changed to _opening a new stream on each retry_ with https://github.com/libp2p/specs/pull/173/commits/5943d3b34a3df937d56c8fa8308a2b5e01922795.